### PR TITLE
Add separate module for the api package

### LINKED ISSUE #1726: Make the API package a separate module
The protobuf descriptors for Dex are currently available in the `api` package which people tend to import in their own application. Unfortunately, this can lead to a number of issues (large number of dependencies, conflicting dependencies, etc) which is why importing application code is usually not advised.

By making the aforementioned package a separate module, people can continue importing it, without importing the entire application at the same time.

This change would be backward compatible, with the exception of multiple modules importing different versions of Dex. Go get would fail in that case. (Microsoft did the same with autorest, it took some time for the community to get used to it)

@bonifaido @srenatus what do you think?

### PR DESCRIPTION
**Pull request by sagikazarmark**
Fixes #1726

This is a first step, we need to need a v2.0.0 version after merging.

### CODE REVIEW

**Comment by sagikazarmark** [comment]
We should probably emphasize in the next release log to import the API only, but not the app module.

**Comment by JoelSpeed** [comment]
Since our vendored modules are checked in, have you run `go mod tidy`, `go mod vendor`, `go mod verify` over this PR? (Do we have a make target for this, we should if not)

**Comment by sagikazarmark** [comment]
Good catch.

Actually: can we get rid of vendor? I really don't see the point (images are built without it, quay is down more than github is).

**Comment by sagikazarmark** [comment]
See #1597

File: go.mod [hunk 1]
```diff
@@ -48,3 +49,5 @@ require (
 	gopkg.in/square/go-jose.v2 v2.4.1
 	sigs.k8s.io/testing_frameworks v0.1.2
 )
+
+replace github.com/dexidp/dex/api => ./api
```
Comment (JoelSpeed) [inline go.mod:53]: Is this necessary, what happens if this is removed? 

Looks like with this, we will have to revendor (if we are keeping that) every time this changes

Comment (sagikazarmark) [reply go.mod:53]: It is necessary. If it is removed, go mod will try to download it from GitHub.

I don't think the api changes that much, but vendoring is not used when building the final image, so no harm done if we don't revendor. :)

**Pull request merged**